resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java

```diff
@@ -889,14 +889,56 @@ protected void restoreMarkers(IResource resource, boolean generateDeltas, IProgr
 			return;
 		}
 		IProject[] projects = ((IWorkspaceRoot) resource).getProjects(IContainer.INCLUDE_HIDDEN);
-		for (IProject project : projects)
-			if (project.isAccessible())
-				markerManager.restore(project, generateDeltas, monitor);
+		runParallel(projects, monitor, project -> {
+			if (project.isAccessible()) {
+				markerManager.restore(project, generateDeltas, null);
+			}
+		});
 		if (Policy.DEBUG_RESTORE_MARKERS) {
 			Policy.debug("Restore Markers for workspace: " + (System.currentTimeMillis() - start) + "ms"); //$NON-NLS-1$ //$NON-NLS-2$
 		}
 	}
 
+	@FunctionalInterface
+	public interface CoreConsumer<T> {
+		/**
+		 * Performs this operation on the given argument.
+		 *
+		 * @param t the input argument
+		 */
+		void accept(T t) throws CoreException;
+	}
+
+	void runParallel(IProject[] projects, IProgressMonitor m, CoreConsumer<IProject> consumer) throws CoreException {
+		// Never use a shared ForkJoinPool.commonPool() as it may be busy with other tasks, which might deadlock.
+		// Also use a custom ForkJoinWorkerThreadFactory, to prevent issues with a
+		// potential SecurityManager, since the threads created by it get no permissions.
+		// See https://github.com/eclipse-platform/eclipse.platform/issues/294
+		ForkJoinPool forkJoinPool = new ForkJoinPool(ForkJoinPool.getCommonPoolParallelism(),
+				pool -> new ForkJoinWorkerThread(pool) {
+					// anonymous subclass to access protected constructor
+				}, null, false);
+		IStatus[] stats;
+		try {
+			stats = forkJoinPool.submit(() -> Arrays.stream(projects).parallel().map(project -> {
+				try {
+					consumer.accept(project);
+				} catch (CoreException e) {
+					return e.getStatus();
+				}
+				return null;
+			}).filter(Objects::nonNull).toArray(IStatus[]::new)).get();
+		} catch (InterruptedException | ExecutionException e) {
+			throw new CoreException(Status.error(Messages.resources_saveProblem, e));
+		} finally {
+			forkJoinPool.shutdown();
+		}
+		if (stats.length > 0) {
```

Comment:
If length == 1 better not throw a multi status

Reply:
Extracases complicate sourcecode for things that should not happen anyway.

Reply:
What will not happen? That only one error occurs? You will find these "complications" all over the Eclipse codebase as Multistatus has a poor rendering in logfiles and dialogs so should be avoided if not needed.

Reply:
That there is an Exception may happen only on error but it does not require a beauty full error log. The case that there is an exactly one error is even more rare.